hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java

```diff
@@ -197,6 +226,36 @@ public HoodieWriteClient(JavaSparkContext jsc, HoodieWriteConfig clientConfig,
     }
   }
 
+  /**
+   * Inserts the given prepared records into the Hoodie table, at the supplied commitTime.
+   *
+   * This implementation skips the index check, skips de-duping and is able to leverage benefits
+   * such as small file handling/blocking alignment, as with insert(), by profiling the workload.
+   * The prepared HoodieRecords should be de-duped if needed.
+   *
+   * @param preppedRecords HoodieRecords to insert
+   * @param commitTime Commit Time handle
+   * @return JavaRDD[WriteStatus] - RDD of WriteStatus to inspect errors and counts
+   */
+  public JavaRDD<WriteStatus> insertPreppedRecords(JavaRDD<HoodieRecord<T>> preppedRecords,
```

Comment:
can't we just call upsertPreppedRecords from here?  the code seems redundant?

Reply:
+1

Reply:
At line 249, we use a different upsertRecordsInternal.
return upsertRecordsInternal(preppedRecords, commitTime, table, false);

upsertPreppedRecords uses:
return upsertRecordsInternal(preppedRecords, commitTime, table, true);

Reply:
sure. I was thinking about the not repeating the other lines.. Should we explore a closure?  My point is previously we reorged everything into upsertRecordsInternal for the same reason and now we are adding 3 more such APIs, its healthy to revisit for code reuse. 

Can we move lines 242-246 into a helper that returns a HoodieTable object.. This can be reused across all the write api methods - including upsert, insert, bulkInsert

Reply:
Looks in `HoodieWriteClient:upsert`, we need HoodieTable for index tagging.
```
      // perform index loop up to get existing location of records
      JavaRDD<HoodieRecord<T>> taggedRecords = index.tagLocation(dedupedRecords, table);
      return upsertRecordsInternal(taggedRecords, commitTime, table, true);
```
Similarly we need to have writeContext = metrics.getCommitCtx() outside of upsertRecordsInternal as this is a timer to measure times spent in APIs like upsert which include three operations:
```
1. dedup - combineOnCondition
2. index tagging - index.tagLocation
3. upsertRecordsInternal
```
or for insert where `writeContext = metrics.getCommitCtx()` needs to be outside of `upsertRecordsInternal`, because it measure times spent in:
```
1. dedup
2. upsertRecordsInternal
```

We might be able to just move `writeContext = metrics.getCommitCtx()` inside of `upsertRecordsInternal` for `insertPreppedRecords` or `upsertPreppedRecords`, but this means that we need another flag in `upsertRecordsInternal` to be able to invoke `writeContext = metrics.getCommitCtx()` if called from `insertPreppedRecords` or `upsertPreppedRecords`.

WDYT?

Reply:
Ya adding `getTableAndInitCtx` can reduce to 1 statement from 2 statements.

Reply:
yes 5 lines to 1 precisely. can we do that and move on here?  

I will reserve long term cleanup for later.